src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/SourceTextExtensions.cs

```diff
@@ -201,4 +201,27 @@ public static bool NonWhitespaceContentEquals(this SourceText source, SourceText
 
         return null;
     }
+
+    public static bool TryGetAbsoluteIndex(this SourceText sourceText, int line, int character, out int absoluteIndex)
+    {
+        absoluteIndex = 0;
+        var lineCount = sourceText.Lines.Count;
+        if (line > lineCount ||
+            (line == lineCount && character > 0))
+        {
+            return false;
+        }
```

Comment:
nit: I would probably write this something like below (easier for my old brain to read and probably fewer conditions to evaluate in common case)

```
            if (line >= lineCount)
            {
                if (line > lineCount || character >0)
                {
                    return false;
                }

                // LSP spec allowed a Range to end one line past the end, and character 0. SourceText does not, so we adjust to the final char position
                absoluteIndex = sourceText.Length;
            }
            else
            {
                absoluteIndex = sourceText.Lines[line].Start + character;
            }
```

Reply:
I'm going to leave this for now, since its just a move, and I find that harder to read 😛

I could be convinced about the less conditions though, so I'll have a go at this in a follow up PR, to clean up more Position and Range stuff. This probably needs more logic anyway - just because the line number is valid, doesn't mean the character position on that line is, and there is no check for line > 0.